src/Middleware/RateLimiting/src/RateLimiterOptions.cs

```diff
@@ -44,5 +36,121 @@ public PartitionedRateLimiter<HttpContext> Limiter
+    {
+        if (policyName == null)
+        {
+            throw new ArgumentNullException(nameof(policyName));
+        }
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException("There already exists a policy with the name {name}");
+        }
+
+        UnactivatedPolicyMap.Add(policyName, new PolicyTypeInfo { PolicyType = typeof(TPolicy), PartitionKeyType = typeof(TPartitionKey) });
+
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a new rate limiting policy with the given name.
+    /// </summary>
+    /// <param name="policyName">The name to be associated with the given <see cref="IRateLimiterPolicy{TPartitionKey}"/>.</param>
+    /// <param name="policy">The <see cref="IRateLimiterPolicy{TPartitionKey}"/> to be applied.</param>
+    public RateLimiterOptions AddPolicy<TPartitionKey>(string policyName, IRateLimiterPolicy<TPartitionKey> policy)
+    {
+        if (policyName == null)
+        {
+            throw new ArgumentNullException(nameof(policyName));
+        }
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException("There already exists a policy with the name {name}");
+        }
+
+        if (policy == null)
+        {
+            throw new ArgumentNullException(nameof(policy));
+        }
+
+        PolicyMap.Add(policyName, new DefaultRateLimiterPolicy(ConvertPartitioner<TPartitionKey>(policy.GetPartition), policy.OnRejected));
+
+        return this;
+    }
+
+    internal RateLimiterOptions InternalAddPolicy(string policyName, Func<HttpContext, RateLimitPartition<DefaultKeyType>> partitioner)
+    {
+        if (policyName == null)
+        {
+            throw new ArgumentNullException(nameof(policyName));
+        }
+
+        if (partitioner == null)
+        {
+            throw new ArgumentNullException(nameof(partitioner));
+        }
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException("There already exists a policy with the name {name}");
+        }
+
+        PolicyMap.Add(policyName, new DefaultRateLimiterPolicy(partitioner, null));
+
+        return this;
+    }
+
+    // Converts a partition<TKey> to a Partition<DefaultKeyType<TKey>> to prevent accidental collisions with the keys we create in the the RateLimiterOptionsExtensions.
+    private static Func<HttpContext, RateLimitPartition<DefaultKeyType>> ConvertPartitioner<TPartitionKey>(Func<HttpContext, RateLimitPartition<TPartitionKey>> partitioner)
+    {
+        return (context =>
+        {
+            RateLimitPartition<TPartitionKey> partition = partitioner(context);
+            return new RateLimitPartition<DefaultKeyType>(new DefaultKeyType<TPartitionKey>(partition.PartitionKey), key => partition.Factory(partition.PartitionKey));
```

Comment:
I think this should be `partition.Factory(key)`

Reply:
`key` is a `DefaultKeyType`, `partition` is a `RateLimitPartition<TPartitionKey>`